### PR TITLE
[BugFix] Fix compatibility problem of storage volume

### PR DESCRIPTION
**Pull request by TszKitLo40**
## Why I'm doing:
If user upgrades from 3.0 and the createTableInfo and createDbInfo is replayed, the image will look like: 
```
{"clazz":"SharedDataStorageVolumeMgr","defaultSVId":"starrocks-qa-test-cloud-data","svToD
bs":{},"svToTables":{"null":[12288,81921,49154,65541]}}
```
If Fe is restarted, the map `storageVolumeToDbs` and `storageVolumeToTables` will be {} and {"null":[12288,81921,49154,65541]}. And this bindings mean there is a storage volume called null maps to some tables and dbs. When user use show create table, unexpected exception will be thrown because a storage volume called null does not exist. And this bindings should not be created.
## What I'm doing:
1. Do not bind storage volume to db and table if storage volume id is null in replay phase.
2. When image is loaded, clear the incorrect mapping which maps "null" to dbs and tables.
Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 1]
```diff
@@ -434,7 +434,9 @@ public void replayCreateDb(CreateDbInfo createDbInfo) {
         try {
             Database db = new Database(createDbInfo.getId(), createDbInfo.getDbName());
             unprotectCreateDb(db);
-            stateMgr.getStorageVolumeMgr().replayBindDbToStorageVolume(createDbInfo.getStorageVolumeId(), db.getId());
+            if (createDbInfo.getStorageVolumeId() != null) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:437]: add comment to explain in which situation, the volumeId can be null.

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 2]
```diff
@@ -2338,7 +2340,7 @@ public void replayCreateTable(CreateTableInfo info) {
             }
         }
 
-        if (table.isCloudNativeTableOrMaterializedView()) {
+        if (table.isCloudNativeTableOrMaterializedView() && info.getStorageVolumeId() != null) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:2343]: is it good enough, is there any other places that calls replay bind db or table storage volume?

File: fe/fe-core/src/main/java/com/starrocks/server/StorageVolumeMgr.java [hunk 3]
```diff
@@ -377,12 +378,25 @@ public void load(SRMetaBlockReader reader)
 
     @Override
     public void gsonPostProcess() throws IOException {
-        for (Map.Entry<String, Set<Long>> entry : storageVolumeToDbs.entrySet()) {
+        // If user upgrades from 3.0 and the createTableInfo and createDbInfo is replayed,
+        // the image will look like: "svToDbs":{"null":[12288,81921,49154,65541,20485]}, "svToTables":{"null":[12288]}
+        // The mapping null to dbs and tables should be removed. These code can be removed when 3.0 is not supported.
+        for (Iterator<Map.Entry<String, Set<Long>>> it = storageVolumeToDbs.entrySet().iterator(); it.hasNext();) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/server/StorageVolumeMgr.java:384]: looks like Iterator is not necessary?

if want to change the `storageVolumeToDbs`, can just remove it directly before the for loop.

File: fe/fe-core/src/test/java/com/starrocks/server/SharedDataStorageVolumeMgrTest.java [hunk 4]
```diff
@@ -870,4 +870,28 @@ public void testCreateStorageVolumeWithInvalidParams() throws DdlException {
                     () -> svm.createStorageVolume(svName, "s3", locations, params, Optional.empty(), ""));
         }
     }
+
+    @Test
+    public void testUpgrade() throws IOException, SRMetaBlockException, SRMetaBlockEOFException,
+            DdlException, AlreadyExistsException {
+        StorageVolumeMgr svm = new SharedDataStorageVolumeMgr();
+        svm.createBuiltinStorageVolume();
+        svm.replayBindDbToStorageVolume(null, 1L);
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/server/SharedDataStorageVolumeMgrTest.java:879]: we should block this call with `null` as parameter.

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java (hunk 2 above)
Comment (TszKitLo40) [reply fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:2345]: yes. I have checked.

File: fe/fe-core/src/main/java/com/starrocks/server/StorageVolumeMgr.java (hunk 3 above)
Comment (TszKitLo40) [reply fe/fe-core/src/main/java/com/starrocks/server/StorageVolumeMgr.java]: done

File: fe/fe-core/src/test/java/com/starrocks/server/SharedDataStorageVolumeMgrTest.java (hunk 4 above)
Comment (TszKitLo40) [reply fe/fe-core/src/test/java/com/starrocks/server/SharedDataStorageVolumeMgrTest.java]: done

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java (hunk 1 above)
Comment (TszKitLo40) [reply fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:438]: done

File: fe/fe-core/src/test/java/com/starrocks/server/SharedDataStorageVolumeMgrTest.java (hunk 4 above)
Comment (TszKitLo40) [reply fe/fe-core/src/test/java/com/starrocks/server/SharedDataStorageVolumeMgrTest.java]: done

**Review by kevincai — DISMISSED** [review]

**Review by HangyuanLiu — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/server/SharedDataStorageVolumeMgr.java [hunk 5]
```diff
@@ -132,6 +132,9 @@ private StorageVolume getStorageVolumeOfDb(String svName) throws DdlException {
     // Because it has been checked when creating db.
     private boolean bindDbToStorageVolume(String svId, long dbId, boolean isReplay) {
         try (LockCloseable lock = new LockCloseable(rwLock.writeLock())) {
+            if (svId == null) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/server/SharedDataStorageVolumeMgr.java:135]: null check should out of the lock. reduce unnecessary lock as much as possible.

Comment (TszKitLo40) [reply fe/fe-core/src/main/java/com/starrocks/server/SharedDataStorageVolumeMgr.java]: done

**Review by kevincai — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**